src/script-child-process.ts

```diff
@@ -113,94 +127,96 @@ export class ScriptChildProcess {
       detached: !IS_WINDOWS,
     });
 
-    this.completed = new Promise((resolve, reject) => {
-      this._child.on('spawn', () => {
-        switch (this._state) {
-          case 'starting': {
-            this._state = 'started';
-            break;
-          }
-          case 'killing': {
-            // We received a kill request while we were still starting. Kill now
-            // that we're started.
-            this._actuallyKill();
-            break;
-          }
-          case 'started':
-          case 'stopped': {
-            reject(
-              new Error(
-                `Internal error: Expected ScriptChildProcessState ` +
-                  `to be "started" or "killing" but was "${this._state}"`
-              )
-            );
-            break;
-          }
-          default: {
-            const never: never = this._state;
-            reject(
-              new Error(
-                `Internal error: unexpected ScriptChildProcessState: ${String(
-                  never
-                )}`
-              )
-            );
-          }
+    this._child.on('spawn', () => {
+      switch (this._state) {
+        case 'starting': {
+          this._started.resolve({ok: true, value: undefined});
+          this._state = 'started';
+          break;
+        }
+        case 'killing': {
+          this._started.resolve({ok: true, value: undefined});
+          // We received a kill request while we were still starting. Kill now
+          // that we're started.
+          this._actuallyKill();
+          break;
         }
-      });
+        case 'started':
+        case 'stopped': {
+          const exception = new Error(
+            `Internal error: Expected ScriptChildProcessState ` +
+              `to be "started" or "killing" but was "${this._state}"`
+          );
+          this._started.reject(exception);
+          this._completed.reject(exception);
+          break;
+        }
+        default: {
+          const never: never = this._state;
+          const exception = new Error(
+            `Internal error: unexpected ScriptChildProcessState: ${String(
+              never
+            )}`
+          );
+          this._started.reject(exception);
+          this._completed.reject(exception);
+        }
+      }
+    });
 
-      this._child.on('error', (error) => {
-        resolve({
+    this._child.on('error', (error) => {
+      const result = {
+        ok: false,
+        error: {
+          script,
+          type: 'failure',
+          reason: 'spawn-error',
+          message: error.message,
+        },
+      } as const;
+      this._started.resolve(result);
+      this._completed.resolve(result);
+      this._state = 'stopped';
+    });
+
+    this._child.on('close', (status, signal) => {
+      if (this._state === 'killing') {
```

Comment:
Do you need to ensure or assert that `_started` is settled in any of these cases?

Reply:
I don't think so. `close` should only happen after `spawn` or `error`.